consensus/src/member.rs

```diff
@@ -107,35 +106,44 @@ enum TaskDetails<H: Hasher, D: Data, S: Signature> {
 }
 
 #[derive(Clone)]
-pub struct LocalIO<
-    D: Data,
-    DP: DataProvider<D>,
-    FH: FinalizationHandler<D>,
-    US: AsyncWrite,
-    UL: AsyncRead,
-> {
+pub struct LocalIO<DP: DataProvider, FH, US: AsyncWrite, UL: AsyncRead> {
     data_provider: DP,
     finalization_handler: FH,
     unit_saver: US,
     unit_loader: UL,
-    _phantom: PhantomData<D>,
 }
 
-impl<D: Data, DP: DataProvider<D>, FH: FinalizationHandler<D>, US: AsyncWrite, UL: AsyncRead>
-    LocalIO<D, DP, FH, US, UL>
+impl<DP: DataProvider, FH: FinalizationHandler<DP::Output>, US: AsyncWrite, UL: AsyncRead>
+    LocalIO<DP, FH, US, UL>
 {
     pub fn new(
         data_provider: DP,
         finalization_handler: FH,
         unit_saver: US,
         unit_loader: UL,
-    ) -> LocalIO<D, DP, FH, US, UL> {
+    ) -> Self {
+        LocalIO {
+            data_provider,
+            finalization_handler,
+            unit_saver,
+            unit_loader,
+        }
+    }
+
+    pub fn new_using_unit_finalization_handler<
+        H: Hasher,
+        UFH: FinalizationHandler<BatchOfUnits<DP::Output, H>>,
+    >(
+        data_provider: DP,
+        finalization_handler: UFH,
+        unit_saver: US,
+        unit_loader: UL,
+    ) -> LocalIO<DP, UFH, US, UL> {
```

Comment:
Returning something other than `Self` here is quite unusual. I think it might be better to have this whole `impl` block be for `LocalIO<DP, UFH, US, UL>`, both the constructors return `Self` and the plain `new` is parametrized by `FH: FinalizationHandler<DP::Output>` (but still returns `Self` – that should work because of the autoimpl, right?).

Or is there some problem with `H` being in the `impl`, but not a parameter of `LocalIO`?